src/python/pants/engine/internals/target_adaptor.py

```diff
@@ -104,10 +104,19 @@ def __init__(
     ) -> None:
         self.type_alias = type_alias
         self.name = name
-        self.kwargs = kwargs
+        self.kwargs = FrozenDict.deep_freeze(kwargs)
         self.description_of_origin = __description_of_origin__
         self.origin_sources_blocks = __origin_sources_blocks__
 
+    def with_new_kwargs(self, **kwargs) -> TargetAdaptor:
+        return TargetAdaptor(
```

Comment:
You can probably use `dataclasses.replace` here?

Reply:
Good idea

Reply:
Oh wait, no, this is not a dataclass...

Reply:
Ack, disregard comment then.